src/libcore/fmt/mod.rs

```diff
@@ -1310,11 +1310,21 @@ impl Display for bool {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl Debug for str {
     fn fmt(&self, f: &mut Formatter) -> Result {
-        try!(write!(f, "\""));
-        for c in self.chars().flat_map(|c| c.escape_default()) {
-            try!(f.write_char(c))
+        try!(f.write_char('"'));
+        let mut from = 0;
+        for (i, c) in self.char_indices() {
+            let esc = c.escape_default();
+            // If char needs escaping, flush backlog so far and write, else skip
+            if esc.size_hint().0 != 1 {
```

Comment:
@alexcrichton Do you mean `esc.size_hint().1 != Some(1)` or more explicit `esc.size_hint() != (1, Some(1))`?

As @bluss I would prefer if we could add method to iterator. That might communicate intention clearer and it might be simpler than `size_hint`.